pdns/pdns_recursor.cc

```diff
@@ -396,114 +396,113 @@ static bool isHandlerThread()
   return s_threadInfos.at(t_id).isHandler;
 }
 
-static void handleTCPClientWritable(int fd, FDMultiplexer::funcparam_t& var);
-
-LWResult::Result asendtcp(const PacketBuffer& data, Socket* sock)
-{
-  PacketID pident;
-  pident.tcpsock=sock->getHandle();
-  pident.outMSG = data;
-
-  t_fdm->addWriteFD(sock->getHandle(), handleTCPClientWritable, pident);
-  PacketBuffer packet;
+#if 0
+#define TCPLOG(x) cerr << [](){ timeval t; gettimeofday(&t, nullptr); return t.tv_sec % 10  + t.tv_usec/1000000.0; }() << ' ' << x
+#else
+#define TCPLOG(x)
+#endif
 
-  int ret = MT->waitEvent(pident, &packet, g_networkTimeoutMsec);
-  if (ret == 0) { //timeout
-    t_fdm->removeWriteFD(sock->getHandle());
-    return LWResult::Result::Timeout;
-  }
-  else if (ret == -1) { // error
-    t_fdm->removeWriteFD(sock->getHandle());
-    return LWResult::Result::PermanentError;
-  }
-  else if (packet.size() != data.size()) { // main loop tells us what it sent out, or empty in case of an error
-    return LWResult::Result::PermanentError;
-  }
+static void TCPIOHandlerIO(int fd, FDMultiplexer::funcparam_t& var);
 
-  return LWResult::Result::Success;
-}
+LWResult::Result asendtcp(const PacketBuffer& data, shared_ptr<TCPIOHandler>& handler)
+{
+  TCPLOG("asendtcp called " << data.size() << endl);
 
-static void TCPIOHandlerWritable(int fd, FDMultiplexer::funcparam_t& var);
+  auto pident = std::make_shared<PacketID>();
+  pident->tcphandler = handler;
+  pident->tcpsock = handler->getDescriptor();
+  pident->outMSG = data;
+  pident->highState = TCPAction::DoingWrite;
+  pident->lowState = IOState::NeedWrite;
 
-LWResult::Result asendtcp(const PacketBuffer& data, TCPIOHandler& handler)
-{
-  PacketID pident;
-  pident.tcphandler = &handler;
-  pident.tcpsock = handler.getDescriptor();
-  pident.outMSG = data;
+  t_fdm->addWriteFD(handler->getDescriptor(), TCPIOHandlerIO, pident);
```

Comment:
I'm afraid writing might require a read, especially if the handshake has not been completed :-/

Reply:
Now I'm confused again... do we need to wait for a readable fd, or do an actual tryRead()? 

I'm asuming the former:

I suppose If the `tryWrite` call is done and returns `NeedRead`, the readable check will be taken care of.
The only case I can see go wrong is that the fd never becomes writeable... so the `tryWrite` is never done.

But if we set `lowState = NeedRead` and call `addReadFD` here, aren't we running the risk that the fd never becomes readable?

Reply:
Yes, the former. If the `tryWrite` call is done and returns `NeedRead`, we need to wait for the fd to become readable then call `tryWrite`. At this point if the is not writable `tryWrite` will return `NeedWrite` and we will have to wait again, for the fd to become writable this time.
If the fd never becomes readable, something is wrong and the timeout will kick in, I think?

Reply:
So my understanding is I should do an explicit tryWrite (without waiting for a condition) to get the flow going, like a mirror of `arcvtcp` that does a `tryRead` first?

Reply:
Yes! That might even save a round-trip when fast open works properly, for example :-)